src/orange/orangeinp/ConvexRegion.cc

```diff
@@ -75,6 +75,178 @@ void Box::build(ConvexSurfaceBuilder& insert_surface) const
+    real_type const tangent = std::fabs(lo - hi) / (2 * hh_);
+
+    // Calculate vanishing point (origin)
+    real_type vanish_z = 0;
+    if (lo > hi)
+    {
+        // Cone opens downward (base is on bottom)
+        vanish_z = -hh_ + lo / tangent;
+        CELER_ASSERT(vanish_z > 0);
+    }
+    else
+    {
+        // Cone opens upward
+        vanish_z = hh_ - hi / tangent;
+        CELER_ASSERT(vanish_z < 0);
+    }
+
+    // Build the cone surface along the given axis
+    ConeZ cone{Real3{0, 0, vanish_z}, tangent};
+    insert_surface(cone);
+
+    // Set radial extents of exterior bbox
+    insert_surface(Sense::inside, make_xyradial_bbox(std::fmax(lo, hi)));
+
+    // Calculating the interior bounding box:
+    // - Represent a radial slice of the cone as a right triangle with base b
+    //   (aka the higher radius) and height h (translated vanishing point)
+    // - An interior bounding box (along the xy diagonal cut!) will satisfy
+    //   r = b - tangent * z
+    // - Maximizing the area of that box gives r = b / 2, i.e. z = h / 2
+    // - Truncate z so that it's not outside of the half-height
+    // - project that radial slice onto the xz plane by multiplying 1/sqrt(2)
+    real_type const b = std::fmax(lo, hi);
+    real_type const h = b / tangent;
+    real_type const z = std::fmin(h / 2, 2 * hh_);
+    real_type const r = b - tangent * z;
+
+    // Now convert from "triangle above z=0" to "cone centered on z=0"
+    auto const [zmin, zmax] = [&]() -> std::pair<real_type, real_type> {
+        if (lo > hi)
+            return {-hh_, -hh_ + z};  // Base is on bottom
+        else
+            return {hh_ - z, hh_};  // Base is on top
+    }();
+    CELER_ASSERT(zmin < zmax);
+    real_type const rbox = (constants::sqrt_two / 2) * r;
+    BBox const interior_bbox{{-rbox, -rbox, zmin}, {rbox, rbox, zmax}};
+
+    // Check that the corners are actually inside the cone
+    CELER_ASSERT(cone.calc_sense(interior_bbox.lower() * real_type(1 - 1e-5))
+                 == SignedSense::inside);
+    CELER_ASSERT(cone.calc_sense(interior_bbox.upper() * real_type(1 - 1e-5))
+                 == SignedSense::inside);
+    insert_surface(Sense::outside, interior_bbox);
+}
+
+//---------------------------------------------------------------------------//
+// CYLINDER
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with radius.
+ */
+Cylinder::Cylinder(real_type radius, real_type halfheight)
+    : radius_{radius}, hh_{halfheight}
+{
+    CELER_VALIDATE(radius_ > 0, << "nonpositive radius: " << radius_);
+    CELER_VALIDATE(hh_ > 0, << "nonpositive half-height: " << hh_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build surfaces.
+ */
+void Cylinder::build(ConvexSurfaceBuilder& insert_surface) const
+{
+    insert_surface(Sense::outside, PlaneZ{-hh_});
+    insert_surface(Sense::inside, PlaneZ{hh_});
+    insert_surface(CCylZ{radius_});
```

Comment:
Why aren't you inserting an interior and exterior bounding box for the cylinder region?

Reply:
Ah, the "known closed" quadric/planar surfaces (axis aligned cylinders, spheres, and planes) automatically generate the correct bounding boxes using the surface clipper. I'll add a note to the convex region builder pointing out the relationship.